Add table metadata service

inv-jishnu opened the pull request:
## Description

Added table metadata service and some related DTO classes

## Related issues and/or PRs

NA

## Changes made

- Added table metadata service and some related DTO classes. 
- The tablemetadata service is also used in data loader API.
- The spotbug exlude is the same as the one added in other PR to resolve issue due Lombok.

## Checklist

> The following is a best-effort checklist. If any items in this checklist are not applicable to this PR or are dependent on other, unmerged PRs, please still mark the checkboxes after you have read and understood each item.

- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes.
- [ ] Any remaining open issues linked to this PR are documented and up-to-date (Jira, GitHub, etc.).
- [ ] Tests (unit, integration, etc.) have been added for the changes.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes in other PRs have been merged and published.

## Additional notes (optional)

Road map to merge remaining data loader core files. Current status

* **General** 
  * Add ScalarDB Dao and related files - https://github.com/scalar-labs/scalardb/pull/2417
  * TableMetadataService(partially replaced by ConsensusUtils): https://github.com/scalar-labs/scalardb/pull/2434

* **Export**
  * Export options and validations: https://github.com/scalar-labs/scalardb/pull/2435
  * ProducerTasks: 1 PR incoming
  
* **Import**
  *  Dto classes and utilities: 1 PR incoming
  * Import processor and task code:  2-3 PRs incoming
  * Code for Import transaction batch and  data chunk import:  1 PR Incoming 
  * ControlFile related Dtos: 1 PR incoming
  * Import logger: 1 PR incoming
  * Import Manager:  1 PR incoming

## Release notes

NA

komamitsu commented on data-loader/core/src/main/java/com/scalar/db/dataloader/core/tablemetadata/TableMetadataRequest.java:
```diff
@@ -0,0 +1,22 @@
+@Getter
+public class TableMetadataRequest {
+
+  private final String namespace;
```
[minor] Same as above

komamitsu commented on data-loader/core/src/main/java/com/scalar/db/dataloader/core/util/TableMetadataUtil.java:
```diff
@@ -0,0 +1,144 @@
+   *
+   * @return A set of predefined metadata column names.
+   */
+  public static Set<String> getMetadataColumns() {
```
How about using `com.scalar.db.transaction.consensuscommit.ConsensusCommitUtils#getTransactionMetaColumns` to avoid duplicated logic?

Some other methods might be replaced with `ConsensusCommitUtils` methods.

brfrn169 commented on data-loader/core/src/main/java/com/scalar/db/dataloader/core/util/TableMetadataUtil.java:
```diff
@@ -0,0 +1,144 @@
+   * @param columnNames A set of all column names in the table.
+   * @return {@code true} if the column is a metadata column; {@code false} otherwise.
+   */
+  public static boolean isMetadataColumn(
```
You can use `ConsensusCommitUtils.isTransactionMetaColumn()`:
https://github.com/scalar-labs/scalardb/blob/98b5e1b737a5ceab9269c28c7ac0991e3a3e13f8/core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitUtils.java#L201

brfrn169 replied:
How about this? Have you made a change for this?

brfrn169 commented on data-loader/core/src/main/java/com/scalar/db/dataloader/core/util/TableMetadataUtil.java:
```diff
@@ -0,0 +1,144 @@
+   * @param tableMetadata The metadata of the table.
+   * @return {@code true} if the column is a metadata column; {@code false} otherwise.
+   */
+  public static boolean isMetadataColumn(String columnName, TableMetadata tableMetadata) {
```
Ditto.

brfrn169 reviewed (commented):
Left a couple comments. Please take a look when you have time!

inv-jishnu commented:
@brfrn169 san, @brfrn169 san,

I have made changes based on your feedback.
Please take a look at this again when you get a chance.
Thank you

komamitsu reviewed (approved):
LGTM! 👍

brfrn169 commented on data-loader/core/src/main/java/com/scalar/db/dataloader/core/tablemetadata/TableMetadataService.java:
```diff
@@ -0,0 +1,75 @@
+@RequiredArgsConstructor
+public class TableMetadataService {
+
+  private static final String ERROR_MISSING_NAMESPACE_OR_TABLE =
+      "Missing namespace or table: %s, %s";
```
Please move this to `CoreError`. Or we might already have the same message for it in `CoreError`. If so, pease use it.

inv-jishnu commented:
@brfrn169 san,

I have removed the code for `isMetadataColumn` and moved the error message to CoreError class.
Please take a look at this again when you get a chance.
Thank you.

brfrn169 commented on data-loader/core/src/main/java/com/scalar/db/dataloader/core/tablemetadata/TableMetadataService.java:
```diff
@@ -0,0 +1,74 @@
+      }
+      return tableMetadata;
+    } catch (ExecutionException e) {
+      throw new TableMetadataException(
+          CoreError.DATA_LOADER_MISSING_NAMESPACE_OR_TABLE.buildMessage(namespace, tableName),
+          e.getCause());
```
It looks like this exception is thrown when retrieving table metadata fails. In that case, is this error message appropriate?

inv-jishnu replied:
Thank you.
I have added another message.

brfrn169 commented on data-loader/core/src/main/java/com/scalar/db/dataloader/core/util/TableMetadataUtil.java:
```diff
@@ -0,0 +1,90 @@
+   * @param tableMetadata The metadata of the ScalarDB table.
+   * @return A map where keys are column names and values are their corresponding {@link DataType}.
+   */
+  public static Map<String, DataType> extractColumnDataTypes(TableMetadata tableMetadata) {
+    Map<String, DataType> definitions = new HashMap<>();
+    for (String columnName : tableMetadata.getColumnNames()) {
+      definitions.put(columnName, tableMetadata.getColumnDataType(columnName));
+    }
+    return definitions;
+  }
```
I just realized it but we can use `TableMetadata.getColumnDataTypes()` for this.

inv-jishnu replied:
I have removed this code and will use TableMetadata.getColumnDataTypes() instead.
Thank you.

brfrn169 reviewed (commented):
Left a couple of comments. Please take a look!

brfrn169 commented on core/src/main/java/com/scalar/db/common/error/CoreError.java:
```diff
@@ -760,6 +760,10 @@ public enum CoreError implements ScalarDbError {
+  DATA_LOADER_MISSING_NAMESPACE_OR_TABLE(
+      Category.USER_ERROR, "0165", "Missing namespace or table: %s, %s", "", ""),
+  DATA_LOADER_TABLE_METADATA_RETRIEVAL_FAILED(
+      Category.USER_ERROR, "0166", "Failed to retrieve table metadata. Details:  %s", "", ""),
```
Nit:
```suggestion
      Category.USER_ERROR, "0166", "Failed to retrieve table metadata. Details: %s", "", ""),
```

brfrn169 commented on data-loader/core/src/main/java/com/scalar/db/dataloader/core/tablemetadata/TableMetadataService.java:
```diff
@@ -0,0 +1,74 @@
+    } catch (ExecutionException e) {
+      throw new TableMetadataException(
+          CoreError.DATA_LOADER_TABLE_METADATA_RETRIEVAL_FAILED.buildMessage(e.getMessage()),
+          e.getCause());
```
Why `e.getCause()` instead of `e`?
```suggestion
          e);
```

inv-jishnu commented:
@brfrn169 san,
I have made changes as suggested.
Please take a look at this again when you get a chance.
Thank you.

Torch3333 commented on core/src/main/java/com/scalar/db/common/error/CoreError.java:
```diff
@@ -760,6 +760,10 @@ public enum CoreError implements ScalarDbError {
       "The underlying-storage data type %s is not supported as the ScalarDB %s data type: %s",
       "",
       ""),
+  DATA_LOADER_MISSING_NAMESPACE_OR_TABLE(
+      Category.USER_ERROR, "0165", "Missing namespace or table: %s, %s", "", ""),
+  DATA_LOADER_TABLE_METADATA_RETRIEVAL_FAILED(
+      Category.USER_ERROR, "0166", "Failed to retrieve table metadata. Details: %s", "", ""),
```
When updating or creating error messages. Please explicitly ask @josh-wong to review them to check the English phrasing.
@josh-wong Please check the error message when you get the chance.

Torch3333 reviewed (approved):
LGTM, thank you.

josh-wong reviewed (approved):
LGTM! Thank you!🙇🏻‍♂️

brfrn169 reviewed (approved):
LGTM! Thank you!
